File: orchestration/run_orchestrator.py
from __future__ import annotations

import argparse
import json
import os
from concurrent.futures import ThreadPoolExecutor, as_completed
from dataclasses import asdict, dataclass
from datetime import datetime, timezone
from pathlib import Path
from typing import Any
from urllib import request as urlrequest
from urllib.error import HTTPError


ROOT_DIR = Path(__file__).resolve().parent
PROMPTS_DIR = ROOT_DIR / "prompts"
RUNS_DIR = ROOT_DIR / "runs"


@dataclass
class AgentResult:
    agent: str
    mode: str
    started_at: str
    ended_at: str
    output: str
    error: str | None = None


def read_text(path: Path) -> str:
    return path.read_text(encoding="utf-8")


def read_json(path: Path) -> dict[str, Any]:
    return json.loads(path.read_text(encoding="utf-8"))


def now_iso() -> str:
    return datetime.now(timezone.utc).isoformat(timespec="seconds").replace("+00:00", "Z")


def validate_task(task: dict[str, Any]) -> None:
    required = ["task_id", "title", "goal", "context", "constraints", "targets", "definition_of_done"]
    missing = [key for key in required if key not in task]
    if missing:
        raise ValueError(f"Task file is missing required keys: {', '.join(missing)}")


def to_json_block(data: Any) -> str:
    return json.dumps(data, ensure_ascii=False, indent=2)


def _get_path(data: Any, dotted: str) -> Any:
    current = data
    for part in dotted.split("."):
        if isinstance(current, list):
            current = current[int(part)]
        elif isinstance(current, dict):
            current = current.get(part)
        else:
            return None
    return current


def call_openai(system_prompt: str, user_prompt: str, model: str, temperature: float) -> str:
    try:
        from openai import OpenAI
    except ImportError as exc:
        raise RuntimeError("openai package is not installed. Run: pip install openai") from exc

    if not os.environ.get("OPENAI_API_KEY"):
        raise RuntimeError("OPENAI_API_KEY is not set.")

    client = OpenAI()
    response = client.responses.create(
        model=model,
        instructions=system_prompt,
        input=user_prompt,
        temperature=temperature,
    )
    return response.output_text


def call_internal_auth(
    *,
    base_url: str,
    auth_path: str,
    llm_path: str,
    login_id: str,
    password: str,
    model: str,
    system_prompt: str,
    user_prompt: str,
    token_field: str,
    response_text_path: str,
) -> str:
    auth_url = f"{base_url.rstrip('/')}{auth_path}"
    llm_url = f"{base_url.rstrip('/')}{llm_path}"

    auth_payload = json.dumps({"login_id": login_id, "password": password}).encode("utf-8")
    auth_req = urlrequest.Request(
        auth_url,
        data=auth_payload,
        headers={"Content-Type": "application/json"},
        method="POST",
    )
    try:
        with urlrequest.urlopen(auth_req, timeout=20) as resp:  # noqa: S310
            auth_json = json.loads(resp.read().decode("utf-8"))
    except HTTPError as exc:  # noqa: PERF203
        body = exc.read().decode("utf-8", errors="ignore")
        raise RuntimeError(f"auth request failed: {exc.code} {body}") from exc

    access_token = _get_path(auth_json, token_field)
    if not access_token:
        raise RuntimeError(f"token field not found: {token_field}")

    llm_payload = json.dumps(
        {
            "model": model,
            "messages": [
                {"role": "system", "content": system_prompt},
                {"role": "user", "content": user_prompt},
            ],
        }
    ).encode("utf-8")
    llm_req = urlrequest.Request(
        llm_url,
        data=llm_payload,
        headers={
            "Content-Type": "application/json",
            "Authorization": f"Bearer {access_token}",
        },
        method="POST",
    )
    try:
        with urlrequest.urlopen(llm_req, timeout=60) as resp:  # noqa: S310
            llm_json = json.loads(resp.read().decode("utf-8"))
    except HTTPError as exc:
        body = exc.read().decode("utf-8", errors="ignore")
        raise RuntimeError(f"llm request failed: {exc.code} {body}") from exc

    text = _get_path(llm_json, response_text_path)
    if not text:
        # fallback paths
        text = _get_path(llm_json, "choices.0.message.content") or llm_json.get("output_text")
    if not text:
        raise RuntimeError(f"response text not found: {response_text_path}")
    return text


def mock_response(agent: str, task: dict[str, Any], planner_output: str | None = None) -> str:
    lines = [
        f"# {agent} output (mock)",
        "",
        f"- task_id: {task['task_id']}",
        f"- title: {task['title']}",
    ]

    if agent == "planner":
        lines.extend(
            [
                "",
                "## Frontend plan",
                "- Define page/component changes and AG Grid impact check.",
                "## Backend plan",
                "- Define API/schema/service changes and transaction order checks.",
                "## QA plan",
                "- Build scenario list from workflow gates and regression scope.",
                "## Merge gates",
                "- validate:grid (if AG Grid in scope), lint, build, and test scenarios.",
            ]
        )
    elif agent in {"frontend", "backend"}:
        lines.extend(
            [
                "",
                "## Inputs",
                "- Planner output received.",
                "",
                "## Proposed execution",
                "- Break work into 2-3 small commits.",
                "- Add validation steps before merge.",
            ]
        )
    else:
        lines.extend(
            [
                "",
                "## Findings",
                "- No blocking issue in mock mode.",
                "",
                "## Go/No-Go",
                "- GO with required validation gates.",
            ]
        )

    if planner_output and agent != "planner":
        lines.extend(["", "## Planner context excerpt", planner_output[:900]])

    return "\n".join(lines)


def run_agent(
    *,
    agent: str,
    mode: str,
    system_prompt: str,
    user_prompt: str,
    model: str,
    temperature: float,
    task: dict[str, Any],
    planner_output: str | None = None,
    internal_config: dict[str, str] | None = None,
) -> AgentResult:
    started_at = now_iso()
    try:
        if mode == "openai":
            output = call_openai(system_prompt, user_prompt, model, temperature)
        elif mode == "internal-auth":
            if not internal_config:
                raise RuntimeError("internal-auth config is missing")
            output = call_internal_auth(
                base_url=internal_config["base_url"],
                auth_path=internal_config["auth_path"],
                llm_path=internal_config["llm_path"],
                login_id=internal_config["login_id"],
                password=internal_config["password"],
                model=model,
                system_prompt=system_prompt,
                user_prompt=user_prompt,
                token_field=internal_config["token_field"],
                response_text_path=internal_config["response_text_path"],
            )
        else:
            output = mock_response(agent, task, planner_output)
        return AgentResult(agent=agent, mode=mode, started_at=started_at, ended_at=now_iso(), output=output)
    except Exception as exc:  # noqa: BLE001
        return AgentResult(
            agent=agent,
            mode=mode,
            started_at=started_at,
            ended_at=now_iso(),
            output="",
            error=str(exc),
        )


def render_user_prompt(task: dict[str, Any], planner_output: str | None = None) -> str:
    chunks = [
        "Task JSON:",
        to_json_block(task),
    ]
    if planner_output:
        chunks.extend(["", "Planner output:", planner_output])
    return "\n".join(chunks)


def write_file(path: Path, content: str) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(content, encoding="utf-8")


def main() -> int:
    parser = argparse.ArgumentParser(description="VIBE-HR multi-agent orchestration scaffold.")
    parser.add_argument("--task-file", required=True, help="Path to task JSON file.")
    parser.add_argument("--mode", choices=["mock", "openai", "internal-auth"], default="mock")
    parser.add_argument("--model", default="gpt-5-mini")
    parser.add_argument("--temperature", type=float, default=0.2)
    parser.add_argument("--max-workers", type=int, default=3)
    parser.add_argument("--base-url", default=os.environ.get("VIBE_BASE_URL", "http://127.0.0.1:8000"))
    parser.add_argument("--auth-path", default=os.environ.get("VIBE_AUTH_PATH", "/api/v1/auth/login"))
    parser.add_argument("--llm-path", default=os.environ.get("VIBE_LLM_PATH", "/api/v1/llm/chat"))
    parser.add_argument("--login-id", default=os.environ.get("VIBE_LOGIN_ID", "admin-local"))
    parser.add_argument("--password", default=os.environ.get("VIBE_PASSWORD", "admin"))
    parser.add_argument("--token-field", default=os.environ.get("VIBE_TOKEN_FIELD", "access_token"))
    parser.add_argument("--response-text-path", default=os.environ.get("VIBE_RESPONSE_TEXT_PATH", "choices.0.message.content"))
    args = parser.parse_args()

    task_path = Path(args.task_file).resolve()
    task = read_json(task_path)
    validate_task(task)

    internal_config = {
        "base_url": args.base_url,
        "auth_path": args.auth_path,
        "llm_path": args.llm_path,
        "login_id": args.login_id,
        "password": args.password,
        "token_field": args.token_field,
        "response_text_path": args.response_text_path,
    }

    run_id = f"{task['task_id']}_{datetime.now(timezone.utc).strftime('%Y%m%d_%H%M%S')}"
    run_dir = RUNS_DIR / run_id
    run_dir.mkdir(parents=True, exist_ok=True)

    planner_system = read_text(PROMPTS_DIR / "planner.system.md")
    planner_result = run_agent(
        agent="planner",
        mode=args.mode,
        system_prompt=planner_system,
        user_prompt=render_user_prompt(task),
        model=args.model,
        temperature=args.temperature,
        task=task,
        internal_config=internal_config,
    )
    write_file(run_dir / "00_planner.md", planner_result.output or f"ERROR: {planner_result.error}")

    worker_map = {
        "frontend": PROMPTS_DIR / "frontend.worker.system.md",
        "backend": PROMPTS_DIR / "backend.worker.system.md",
        "qa": PROMPTS_DIR / "qa.reviewer.system.md",
    }

    worker_results: dict[str, AgentResult] = {}
    with ThreadPoolExecutor(max_workers=args.max_workers) as executor:
        future_map = {}
        for agent, prompt_path in worker_map.items():
            system_prompt = read_text(prompt_path)
            future = executor.submit(
                run_agent,
                agent=agent,
                mode=args.mode,
                system_prompt=system_prompt,
                user_prompt=render_user_prompt(task, planner_result.output),
                model=args.model,
                temperature=args.temperature,
                task=task,
                planner_output=planner_result.output,
                internal_config=internal_config,
            )
            future_map[future] = agent

        for future in as_completed(future_map):
            result = future.result()
            worker_results[result.agent] = result

    order = [("frontend", "10_frontend.md"), ("backend", "11_backend.md"), ("qa", "12_qa.md")]
    for agent, filename in order:
        result = worker_results.get(agent)
        if result is None:
            write_file(run_dir / filename, "ERROR: missing result")
            continue
        write_file(run_dir / filename, result.output or f"ERROR: {result.error}")

    reviewer_system = read_text(PROMPTS_DIR / "qa.reviewer.system.md")
    reviewer_input = {
        "task": task,
        "planner": planner_result.output,
        "frontend": worker_results.get("frontend").output if worker_results.get("frontend") else "",
        "backend": worker_results.get("backend").output if worker_results.get("backend") else "",
        "qa": worker_results.get("qa").output if worker_results.get("qa") else "",
    }
    reviewer_result = run_agent(
        agent="reviewer",
        mode=args.mode,
        system_prompt=reviewer_system,
        user_prompt=to_json_block(reviewer_input),
        model=args.model,
        temperature=args.temperature,
        task=task,
        internal_config=internal_config,
    )
    write_file(run_dir / "90_reviewer.md", reviewer_result.output or f"ERROR: {reviewer_result.error}")

    summary = {
        "run_id": run_id,
        "task_file": str(task_path),
        "mode": args.mode,
        "model": args.model,
        "results": {
            "planner": asdict(planner_result),
            "frontend": asdict(worker_results.get("frontend")) if worker_results.get("frontend") else None,
            "backend": asdict(worker_results.get("backend")) if worker_results.get("backend") else None,
            "qa": asdict(worker_results.get("qa")) if worker_results.get("qa") else None,
            "reviewer": asdict(reviewer_result),
        },
    }
    write_file(run_dir / "summary.json", json.dumps(summary, ensure_ascii=False, indent=2))

    print(f"Run completed: {run_dir}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: orchestration/run_orchestrator.js
#!/usr/bin/env node
"use strict";

const fs = require("fs");
const path = require("path");

const ROOT_DIR = __dirname;
const PROMPTS_DIR = path.join(ROOT_DIR, "prompts");
const RUNS_DIR = path.join(ROOT_DIR, "runs");
const REQUIRED_TASK_FIELDS = ["task_id", "title", "goal", "context", "constraints", "targets", "definition_of_done"];

function nowIso() { return new Date().toISOString().replace(/\.\d{3}Z$/, "Z"); }
function readText(file) { return fs.readFileSync(file, "utf8"); }
function readJson(file) { return JSON.parse(readText(file)); }
function writeFile(file, value) { fs.mkdirSync(path.dirname(file), { recursive: true }); fs.writeFileSync(file, value, "utf8"); }
function validateTask(task) {
  const missing = REQUIRED_TASK_FIELDS.filter((key) => !(key in task));
  if (missing.length) throw new Error(`Task file is missing required keys: ${missing.join(", ")}`);
  const legacyBackendTargets = Array.isArray(task.targets?.backend)
    ? task.targets.backend.filter((target) => typeof target === "string" && /^backend(?:[\\/]|$)/.test(target))
    : [];
  if (legacyBackendTargets.length) throw new Error(`Task file uses retired backend targets: ${legacyBackendTargets.join(", ")}`);
}
function json(data) { return JSON.stringify(data, null, 2); }
function getPath(data, dotted) {
  return dotted.split(".").reduce((current, part) => {
    if (Array.isArray(current)) return /^\d+$/.test(part) ? current[Number(part)] : undefined;
    return current && typeof current === "object" ? current[part] : undefined;
  }, data);
}
function extractText(data, responseTextPath) {
  const candidates = [getPath(data, responseTextPath), getPath(data, "output_text"), getPath(data, "choices.0.message.content"), getPath(data, "result.text"), getPath(data, "message.content")];
  for (const value of candidates) {
    if (typeof value === "string" && value.trim()) return value;
    if (Array.isArray(value)) {
      const joined = value.map((item) => item && typeof item === "object" ? item.text || "" : "").join(" ").trim();
      if (joined) return joined;
    }
  }
  throw new Error(`response text not found: ${responseTextPath}`);
}
async function postJson(url, payload, headers, timeoutSeconds, fetchImpl = fetch) {
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), timeoutSeconds * 1000);
  try {
    const response = await fetchImpl(url, { method: "POST", headers, body: JSON.stringify(payload), signal: controller.signal });
    const body = await response.text();
    if (!response.ok) throw new Error(`${url} failed: HTTP ${response.status} ${body}`);
    try { return JSON.parse(body); } catch { throw new Error(`${url} returned non-JSON response: ${body.slice(0, 500)}`); }
  } finally { clearTimeout(timeout); }
}
async function callOpenAi(systemPrompt, userPrompt, model, temperature, options) {
  if (!options.apiKey) throw new Error("OPENAI_API_KEY is not set.");
  const response = await postJson("https://api.openai.com/v1/responses", { model, instructions: systemPrompt, input: userPrompt, temperature }, { "Content-Type": "application/json", Authorization: `Bearer ${options.apiKey}` }, options.llmTimeoutSec, options.fetchImpl);
  return extractText(response, "output_text");
}
async function buildInternalToken(options) {
  const response = await postJson(options.authUrl, { login_id: options.loginId, password: options.password }, { "Content-Type": "application/json" }, options.authTimeoutSec, options.fetchImpl);
  const token = getPath(response, options.tokenField);
  if (typeof token !== "string" || !token) throw new Error(`token field not found: ${options.tokenField}`);
  return token;
}
async function callInternalAuth(systemPrompt, userPrompt, model, temperature, options) {
  const payload = options.llmBodyStyle === "responses" ? { model, instructions: systemPrompt, input: userPrompt, temperature } : { model, messages: [{ role: "system", content: systemPrompt }, { role: "user", content: userPrompt }], temperature };
  const response = await postJson(options.llmUrl, payload, { "Content-Type": "application/json", Authorization: `Bearer ${options.accessToken}` }, options.llmTimeoutSec, options.fetchImpl);
  return extractText(response, options.responseTextPath);
}
function mockResponse(agent, task, plannerOutput) {
  const sections = [`# ${agent} output (mock)`, "", `- task_id: ${task.task_id}`, `- title: ${task.title}`, ""];
  if (agent === "planner") sections.push("## Frontend plan", "- Define page/component changes and AG Grid impact check.", "## Backend plan", "- Define Spring API/schema/service changes and transaction order checks.", "## QA plan", "- Build scenario list from workflow gates and regression scope.", "## Merge gates", "- validate:grid (if AG Grid in scope), Gradle tests, lint, build, and test scenarios.");
  else if (["frontend", "backend"].includes(agent)) sections.push("## Inputs", "- Planner output received.", "", "## Proposed execution", "- Break work into 2-3 small commits.", "- Add validation steps before merge.");
  else sections.push("## Findings", "- No blocking issue in mock mode.", "", "## Go/No-Go", "- GO with required validation gates.");
  if (plannerOutput && agent !== "planner") sections.push("", "## Planner context excerpt", plannerOutput.slice(0, 900));
  return sections.join("\n");
}
function userPrompt(task, plannerOutput) { return ["Task JSON:", json(task), ...(plannerOutput ? ["", "Planner output:", plannerOutput] : [])].join("\n"); }
async function runAgent(agent, mode, systemPrompt, prompt, options, plannerOutput) {
  const started_at = nowIso();
  try {
    const output = mode === "openai" ? await callOpenAi(systemPrompt, prompt, options.model, options.temperature, options)
      : mode === "internal-auth" ? await callInternalAuth(systemPrompt, prompt, options.model, options.temperature, options)
        : mockResponse(agent, options.task, plannerOutput);
    return { agent, mode, started_at, ended_at: nowIso(), output, error: null };
  } catch (error) { return { agent, mode, started_at, ended_at: nowIso(), output: "", error: error.message }; }
}
function joinBaseAndPath(baseUrl, endpointPath) { return `${baseUrl.replace(/\/+$/, "")}/${endpointPath.replace(/^\/+/, "")}`; }
function parseArgs(argumentsList, env = process.env) {
  const options = { taskFile: "", mode: "mock", model: "gpt-5-mini", temperature: 0.2, maxWorkers: 3, baseUrl: env.VIBE_BASE_URL || "http://127.0.0.1:8080", authUrl: env.VIBE_AUTH_URL || "", authPath: env.VIBE_AUTH_PATH || "/api/v1/auth/login", llmUrl: env.VIBE_LLM_URL || "", llmPath: env.VIBE_LLM_PATH || "/api/v1/llm/chat", loginId: env.VIBE_LOGIN_ID || "", password: env.VIBE_PASSWORD || "", accessToken: env.VIBE_ACCESS_TOKEN || "", tokenField: env.VIBE_TOKEN_FIELD || "access_token", responseTextPath: env.VIBE_RESPONSE_TEXT_PATH || "choices.0.message.content", authTimeoutSec: Number(env.VIBE_AUTH_TIMEOUT_SEC || 20), llmTimeoutSec: Number(env.VIBE_LLM_TIMEOUT_SEC || 60), llmBodyStyle: env.VIBE_LLM_BODY_STYLE || "chat", apiKey: env.OPENAI_API_KEY || "", fetchImpl: fetch };
  const fields = { "--task-file": "taskFile", "--mode": "mode", "--model": "model", "--temperature": "temperature", "--max-workers": "maxWorkers", "--base-url": "baseUrl", "--auth-url": "authUrl", "--auth-path": "authPath", "--llm-url": "llmUrl", "--llm-path": "llmPath", "--login-id": "loginId", "--password": "password", "--access-token": "accessToken", "--token-field": "tokenField", "--response-text-path": "responseTextPath", "--auth-timeout-sec": "authTimeoutSec", "--llm-timeout-sec": "llmTimeoutSec", "--llm-body-style": "llmBodyStyle" };
  for (let index = 0; index < argumentsList.length; index += 1) {
    const argument = argumentsList[index]; const field = fields[argument];
    if (!field || !argumentsList[index + 1]) throw new Error(`Unknown or incomplete argument: ${argument}`);
    options[field] = argumentsList[++index];
  }
  options.temperature = Number(options.temperature); options.maxWorkers = Number(options.maxWorkers); options.authTimeoutSec = Number(options.authTimeoutSec); options.llmTimeoutSec = Number(options.llmTimeoutSec);
  if (!options.taskFile || !["mock", "openai", "internal-auth"].includes(options.mode) || !["chat", "responses"].includes(options.llmBodyStyle) || !Number.isInteger(options.maxWorkers) || options.maxWorkers < 1) throw new Error("Invalid required options.");
  return options;
}
async function runWorkers(workers, options, plannerOutput) {
  const results = [];
  const queue = [...workers];
  const workerCount = Math.min(options.maxWorkers, queue.length);
  await Promise.all(Array.from({ length: workerCount }, async () => {
    while (queue.length) {
      const [agent, promptFile] = queue.shift();
      results.push(await runAgent(agent, options.mode, readText(path.join(PROMPTS_DIR, promptFile)), userPrompt(options.task, plannerOutput), options, plannerOutput));
    }
  }));
  return results;
}
async function orchestrate(options) {
  const taskFile = path.resolve(options.taskFile); const task = readJson(taskFile); validateTask(task); options.task = task;
  if (options.mode === "internal-auth") {
    options.authUrl = options.authUrl.trim() || joinBaseAndPath(options.baseUrl, options.authPath);
    options.llmUrl = options.llmUrl.trim() || joinBaseAndPath(options.baseUrl, options.llmPath);
    if (!options.accessToken.trim()) {
      if (!options.loginId.trim() || !options.password) throw new Error("internal-auth mode requires either --access-token or both --login-id/--password.");
      options.accessToken = await buildInternalToken(options);
    }
  }
  const runId = `${task.task_id}_${new Date().toISOString().replace(/[-:]/g, "").replace(/\.\d+Z$/, "")}`; const runDir = path.join(options.runsDir || RUNS_DIR, runId); fs.mkdirSync(runDir, { recursive: true });
  const planner = await runAgent("planner", options.mode, readText(path.join(PROMPTS_DIR, "planner.system.md")), userPrompt(task), options); writeFile(path.join(runDir, "00_planner.md"), planner.output || `ERROR: ${planner.error}`);
  const workers = [["frontend", "frontend.worker.system.md", "10_frontend.md"], ["backend", "backend.worker.system.md", "11_backend.md"], ["qa", "qa.reviewer.system.md", "12_qa.md"]];
  const workerResults = await runWorkers(workers, options, planner.output);
  const byAgent = Object.fromEntries(workerResults.map((result) => [result.agent, result]));
  for (const [agent, , outputFile] of workers) { const result = byAgent[agent]; writeFile(path.join(runDir, outputFile), result?.output || `ERROR: ${result?.error || "missing result"}`); }
  const reviewerInput = { task, planner: planner.output, frontend: byAgent.frontend?.output || "", backend: byAgent.backend?.output || "", qa: byAgent.qa?.output || "" };
  const reviewer = await runAgent("reviewer", options.mode, readText(path.join(PROMPTS_DIR, "qa.reviewer.system.md")), json(reviewerInput), options); writeFile(path.join(runDir, "90_reviewer.md"), reviewer.output || `ERROR: ${reviewer.error}`);
  const summary = { run_id: runId, task_file: taskFile, mode: options.mode, model: options.model, results: { planner, frontend: byAgent.frontend || null, backend: byAgent.backend || null, qa: byAgent.qa || null, reviewer } };
  writeFile(path.join(runDir, "summary.json"), json(summary));
  return { runDir, hasError: [planner, ...workerResults, reviewer].some((result) => result.error), summary };
}
async function main() { try { const result = await orchestrate(parseArgs(process.argv.slice(2))); console.log(result.hasError ? `Run completed with errors: ${result.runDir}` : `Run completed: ${result.runDir}`); process.exitCode = result.hasError ? 2 : 0; } catch (error) { console.error(`Startup failed: ${error.message}`); process.exitCode = 2; } }
if (require.main === module) main();
module.exports = { extractText, getPath, joinBaseAndPath, mockResponse, orchestrate, parseArgs, postJson, runWorkers, validateTask };
